content/python/concepts/decorators/decorators.md

```diff
@@ -0,0 +1,482 @@
+---
+Title: 'Decorators'
+Description: '_Decorators_ are a structural design pattern which wrap the original object, to add extra functionality dynamically, without modifying the original object. Decorators are usually used to run code before and/or after a function. Thus, they decorate the function with additional functionality, without changing the original function. They can also be used to alter the functionality of methods and classes.'
+Subjects:
+  - 'Interview Prep'
+  - 'Web Development'
+  - 'Developer Tools'
+Tags:
+  - 'Best Practices'
+  - 'Classes'
+  - 'Decorators'
+  - 'Functions'
+  - 'Objects'
+  - 'Technical Interviews'
+CatalogContent:
+  - 'learn-intermediate-python-3'
+---
+
+_Decorators_ are a structural design pattern which wrap the original object, to add extra functionality dynamically, without modifying the original object. Decorators are usually used to run code before and/or after a function. Thus, they decorate the function with additional functionality, without changing the original function. They can also be used to alter the functionality of methods and classes.
+
+## Creating Decorators
+
+Functions in Python are first class objects: they can be assigned to variables, passed into functions as arguments, returned from functions and declared inside functions.
+
+With that in mind, a decorator can be created using functions or classes.
+
+### Decorator using functions
+
+The general syntax for creating a decorator using functions is as follows:
+
+```py
+def decorator_name(original_function):
+
+    def wrapper():
+        # code can be executed before the original function call
+
+        original_function()
+
+        # code can be executed after the original function call
+
+    # (IMPORTANT)
+    return wrapper
+
+```
+
+The `original_function()` in this case, is the function that is decorated with the `decorator_name()`.
+
+Eg: A decorator that upper cases the returned values from any function can be created as follows:
+
+```codebyte/py
+# define a function called upper_case_decorator which will act as a decorator with parameter original_function
+
+def upper_case_decorator(original_function):
+
+    # creating the wrapper to wrap additional code around the original function
+    def wrapper():
+
+        # Code can be executed before the original function
+        print(f"'{original_function.__name__}' function execution will start now.")
+
+        # Original function call with the upper cased argument
+        result = original_function(arg1)
+
+        # Code can be executed after the original function
+        result = result.upper()
+        return result
+
+    # returning the wrapper function
```

Comment:
```suggestion
    # Return the wrapper function
```

Reply:
@Goku-kun There seems to be still a few comments that need to start with a capital letter. Please could we update these